Implementation of runtime dynamic linking for VPI/VHPI on Windows

tpambor opened the pull request:
This PR implements runtime dynamic linking for VPI libraries on Windows.

Potential benefits have already been outlined in #1390. 

`vpi_*/vhpi_*` functions are resolved only once on their first call and then execution is forwarded to the respective implementation. `vpi_*/vhpi_*` wrappers are implemented thread-safe.

Currently, the compile-time dependency on the DLL name still exists to follow closely how symbols are resolved using the old import library approach. The dependency on the DLL name could be dropped by searching in all modules of the current process for `vpi_*/vhpi_*` functions. This works as long as only one module in the process exposes `vpi_*/vhpi_*` as otherwise it would be ambigious which `vpi_*/vhpi_*` function to use. Therefore, vpi libraries are changed to not expose the internal `vpi_*/vhpi_*` wrappers. Searching all modules is left for a follow-up PR to keep this PR smaller.

The `endian_swapper` has been run 10 times to evaluate the performance effect of this PR.
```
Without this PR: SIM / REAL TIME 25227 +- 795 NS/S
With this PR:    SIM / REAL TIME 25156 +- 968 NS/S
```
No significant performance drawback could be observed.

eric-wieser commented on cocotb/share/def/modelsim.def:
```diff
@@ -1,115 +0,0 @@
-mti_GetSignalSubelements
-mti_GetSignalType
-mti_GetSignalValue
-mti_GetSignalValueIndirect
-mti_GetTopRegion
-mti_GetTypeKind
-mti_GetVarKind
-mti_GetVarName
-mti_GetVarSubelements
-mti_GetVarType
-mti_GetVarValue
-mti_GetVarValueIndirect
-mti_NextRegion
-mti_NextSignal
-mti_NextVar
-mti_Now
-mti_NowUpper
-mti_Quit
-mti_ScheduleWakeup
-mti_ScheduleWakeup64
-mti_Sensitize
-mti_SetSignalValue
-mti_SetVarValue
-mti_TickLeft
-mti_TickLength
-mti_TickRight
-mti_VsimFree
-mti_Interp
-mti_Cmd
-Tcl_GetStringResult
-Tcl_ResetResult
-Tcl_GetObjResult
-Tcl_ResetResult
-Tcl_ListObjGetElements
-Tcl_GetStringResult
-TclFreeObj
-Tcl_ResetResult
-Tcl_ResetResult
-Tcl_GetString
-TclFreeObj
```
What about all these non-vpi bits?

tpambor replied:
These are loaded from the import library shipped with modelsim `mtipli.lib`. We already dependent on the header files shipped with modelsim so this does not introduce any additional dependency

themperek replied:
I will try to chek but have filling that it has to be done for fli and vhpi (tcl?) also to work without def files. 

We also remove simple def file and add more c code? Is there more advantage than this?

eric-wieser replied:
Agreed, it's not clear to me what the benefit of this is.

tpambor replied:
I have tested it, VPI and FLI for modelsim compile fine. If VHPI is also converted to runtime dynamic linking, all .def files could be dropped and the path to compile with MSVC would be unblocked. Actually, if uncommenting the aldec vhpi library and applying a bunch of hacks (adapting the compiler flags for MSVC, adding the __declspec(dllexport) where necesarry), cocotb can be installed with `pip install .` using the official binary version of python for windows.

Moreover, if the dependency on the DLL name is dropped, the `libcocotbvpi_*` could be easily merged into a single `libcocotbvpi` library.

tpambor replied:
Additionally, icarus master, starting with [`a59b183b`](https://github.com/steveicarus/iverilog/commit/a59b183bd1890eff63cc59d38084dde15de281ff#diff-9bb3b161b30ef5d95427f8ed62cb9e88R685) is no longer working on windows because instead of exporting `vpi_*` in `vvp.exe`, now a function table is passed. As a result, the current `libcocotbvpi_icarus.vpl` is unable to load because it no longer can import the symbols. Based on this PR, `libcocotbvpi_icarus.vpl` could be easily adapted to support both versions of iverilog.

eric-wieser replied:
> Additionally, icarus master, starting with [`a59b183b`](https://github.com/steveicarus/iverilog/commit/a59b183bd1890eff63cc59d38084dde15de281ff#diff-9bb3b161b30ef5d95427f8ed62cb9e88R685) is no longer working on windows because instead of exporting `vpi_*` in `vvp.exe`, now a function table is passed. As a result, the current `libcocotbvpi_icarus.vpl` is unable to load because it no longer can import the symbols. Based on this PR, `libcocotbvpi_icarus.vpl` could be easily adapted to support both versions of iverilog.

Could you create a new issue to explain and track this?

themperek replied:
>Additionally, icarus master, starting with a59b183b is no longer working on windows

This is a good reason to change.

>Moreover, if the dependency on the DLL name is dropped, the libcocotbvpi_* could be easily merged into a single libcocotbvpi library.

This is may hard to do for other/compatibility reasons, unfortunately.

themperek replied:
>I have tested it, VPI and FLI for modelsim compile fine. If VHPI is also converted to runtime dynamic linking, all .def files could be dropped and the path to compile with MSVC would be unblocked. Actually, if uncommenting the aldec vhpi library and applying a bunch of hacks (adapting the compiler flags for MSVC, adding the __declspec(dllexport) where necesarry), cocotb can be installed with pip install . using the official binary version of python for windows.

Do you know if such dll does actually work/loads without issues?

eric-wieser replied:
Are def files not compatible with msvc? I thought they were.

tpambor replied:
> Are def files not compatible with msvc? I thought they were.

It would be possible. `dlltool` is not part of the msvc toolchain and therefore would need to be substituted. Probably, it would also be necessary to locate the msvc toolchain because typically it is not on `PATH`. IMHO the extra maintenance burden of this way would object it. Whereas this PR reduces dependencies instead of introducing new ones. VPI/VHPI APIs have not really changed in years and therefore I do not expect much maintenance effort in having a wrapper for runtime loading.

eric-wieser replied:
Right, the MSVC equivalent is `lib /def:modelsim.def /out:modelsim.lib /machine:x86`

themperek commented on cocotb_build_libs.py:
```diff
@@ -363,24 +368,37 @@ def get_ext():
-                extra_link_args=_extra_link_args(lib_name=lib_name, rpaths=["$ORIGIN"]),
-                extra_compile_args=_extra_cxx_compile_args,
-            )
+            if _can_load_library(modelsim_libmtipli_path):
```
Not sure if `_can_load_library` is good idea. If I am not mistaken (I can easily be) we do not need to load libraries here (on POSIX this would mean `libmtipli.so` is available to load which may not be the case at this point on Windows `modelsim_lib_path` is on `PATH`). If `mti.h` is available we can probably assume "dll/so" also. Alternative just check `os.path.isfile`?

tpambor replied:
This is not a about checking whether the file is present but rather if it is compatible. If 32-bit modelsim is installed and a 64-bit python environment is used `pip install` results in a linker error because the 32-bit `libmtipli` cannot be linked with the 64-bit `libcocotbfli_modelsim`. Rather than stoping with a linker error I prefer to throw a warning and skip `libcocotbfli_modelsim`. `libmtipli.so` is loaded from the folder which also contains `vdbg`. An absolute path is used and therefore it should not be necessary to have different `LD_LIBRARY_PATH`. It works for me on Windows and Linux.

themperek replied:
OK now I get it. In this case probbaly a line of comment would be good.

themperek commented:
Some very basic crosscheck on windows shows that it is working fine. We will do more if we are ready to merge.

cmarqu commented on cocotb/share/lib/vpi/VpiTrampoline.cpp:
```diff
@@ -0,0 +1,251 @@
+/******************************************************************************
+* Copyright (c) 2020 OSA Systeme GmbH
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*    * Redistributions of source code must retain the above copyright
+*      notice, this list of conditions and the following disclaimer.
+*    * Redistributions in binary form must reproduce the above copyright
+*      notice, this list of conditions and the following disclaimer in the
+*      documentation and/or other materials provided with the distribution.
+*    * Neither the name of OSA Systeme GmbH not the
+*      names of its contributors may be used to endorse or promote products
+*      derived from this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL POTENTIAL VENTURES LTD BE LIABLE FOR ANY
+* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************************************************************/
```
This should be
```
# Copyright cocotb contributors
# Licensed under the Revised BSD License, see LICENSE for details.
# SPDX-License-Identifier: BSD-3-Clause
```
for new files (see ``CONTRIBUTING.md``).

imphil commented:
@tpambor Thanks a lot for this PR, what a great initial cocotb contribution! I'll have a closer look in the coming days, but I certainly appreciate the work to get rid of the tool-specific `def` files, and removing the dependency on the tool being present at installation time.

tpambor commented:
I am going to rebase this PR on #1796 to clean it up a little more

tpambor commented:
Changes:
- Rebased on #1796
- Squashed commetics, cleanup and SPDX license header changes
- Added comment with explanation for `_can_load_library`
- No longer touches `vpi_user.h`, `vhpi_user.h` to keep the version from the standard

eric-wieser commented on cocotb/share/lib/vpi/VpiTrampoline.cpp:
```diff
@@ -0,0 +1,228 @@
+
+        static void *resolve_function(const char *name)
+        {
+#if defined(ALDEC)
+            static HMODULE hModule = GetModuleHandle("aldecpli.dll");
+#elif defined(GHDL)
+            static HMODULE hModule = GetModuleHandle("libghdlvpi.dll");
+#elif defined(ICARUS)
+            static HMODULE hModule = GetModuleHandle("vvp.exe");
+#elif defined(MODELSIM)
+            static HMODULE hModule = GetModuleHandle("mtipli.dll");
+#else
+#error No target module defined for trampoline
+#endif
```
Is this any better than having the names in the `.def` files?

tpambor replied:
This is here for compatibility to not break any of the existing tools. The symbols are resolved here in the same way as the OS does. In the future the module names can be removed by searching for `vpi_*` in the whole process space but this works only if there is only one module exposing `vpi_*` as otherwise it would be ambiguous which to select. As a consequence this requires testing in all tools and is therefore left for a future PR.

imphil replied:
Can you add this piece of information to the function as comment please?

tpambor committed:
Switch to resolving VPI functions at runtime on Windows

tpambor committed:
Use the library shipped with Modelsim instead of creating our own import library

tpambor committed:
Workaround broken portability helpers in vpi_user.h

tpambor committed:
Workaround broken portability helpers in vhpi_user.h

tpambor committed:
Only compile FLI interface if the installed modelsim version is compatible with the current python runtime e.g. 32-bit modelsim and 32-bit python

tpambor committed:
Implement run-time dynamic linking for VHPI

themperek commented:
Since Icarus 11 is released it would be important to put this in.

@eric-wieser Could you have a look?

imphil commented on cocotb/share/lib/vpi/VpiTrampoline.cpp:
```diff
@@ -0,0 +1,228 @@
+#include <cinttypes>
+#include <Windows.h>
+
+/* Define function pointers for routines */
```
Can you expand this comment a bit explaining what this file actually does? It's a rather niche setup we have here, and I'm sure many people will be confused why we have such an "interesting" machinery in here. (Or add more documentation to the `VpiTrampoline` class below.)

imphil commented on cocotb_build_libs.py:
```diff
@@ -495,29 +508,45 @@ def get_ext():
-                extra_compile_args=_extra_cxx_compile_args,
-            )
+            # Check if the library is compatible with our current architecture
+            # e.g. 32-bit libmtipli and 32-bit python runtime
```
python -> Python

imphil commented on cocotb_build_libs.py:
```diff
@@ -495,29 +508,45 @@ def get_ext():
-            )
+            # Check if the library is compatible with our current architecture
+            # e.g. 32-bit libmtipli and 32-bit python runtime
+            # To check can load the library into our current python runtime
+            # if it works we are sure that it is compatible
```
Small wording nit: "To check if the FLI library can be loaded, we simply load it. This catches cases where there is a mismatch between the library architecture and the Python architecture (i.e. a 32 bit FLI library used on 64 bit Python)."

imphil commented on cocotb_build_libs.py:
```diff
@@ -495,29 +508,45 @@ def get_ext():
+                logger.warning(
+                    "Cannot build FLI interface for Modelsim/Questa - "
+                    "the mtipli library for '{}' at '{}' cannot be loaded.".format(
+                        vsim_path, modelsim_libmtipli_path
```
Is there more information in what is actually going wrong in the OSError that you catch in _can_load_library?

imphil commented:
So here we go ... Thanks @tpambor for your PR and the massive amount of work you put into this PR. As expected, the solution isn't a piece of beauty, but it looks like it could do the trick. It's a shame I didn't have time to look into this when you first opened this PR, but better late than never!

Overall, this PR looks good and I'm ready to merge it without major changes.

In addition to the comments inline, some high-level points:
* The C/C++ code has some really long lines and other formatting issues where it doesn't match our typical style. Instead of manually formatting them right now, I'd propose to get this PR in and then do a full clang-format run on all of our C/C++ code. I filed https://github.com/cocotb/cocotb/issues/2189 to keep track of that.
* The duplication of function declarations/definitions in the trampolines is unfortunate. We should look into macro magic (including a file multiple times with right helper macros set), or a generator script. This can be done in a follow-up, though.
* Can you please rename the VpiTrampoline and VhpiTrampoline files to have Windows in their name, e.g. VpiTrampolineWindows?

garmin-mjames commented on cocotb/share/lib/vhpi/VhpiImpl.h:
```diff
@@ -35,9 +35,32 @@
+#if !defined(__linux__) && !defined(__APPLE__)
+#define PLI_DLLISPEC
+
+// Workaround broken WIN32 detection in vhpi_user.h
```
xref #1759
